### PR TITLE
Dima

### PR DESCRIPTION
**Pull request by dbabyuk**
Will be glad to hear any comments and suggestions

### CODE REVIEW

File: locators.py [hunk 1]
```diff
@@ -1,38 +1,30 @@
...
-        'email_id': 'email',
-        'phone_id': 'phoneNumber',
-        'country_id': 'country',
-        'city_id': 'city',
-        'street_id': 'street',
-        'building_id': 'building',
-        'apartment_id': 'apartment',
-        'zipcode_id': 'zipCode'
-        }
+class Locators(object):
```
Comment (TanyaKalistruk) [inline locators.py]: please, change the name of the class, for better understanding

File: methods.py [hunk 2]
```diff
@@ -1,33 +1,38 @@
...
+        """Returns all elements for the specific locator"""
+        return self.driver.find_elements(*locator)
+
+    def click_element(self, locator, elem_number=0):
+        """Clicks on the element with number elem_number"""
+        WebDriverWait(self.driver, 10).until(EC.element_to_be_clickable(locator))
+        elements = self.get_elements(locator)
+        elements[elem_number].click()
+
+    def click_element_by_text(self, locator, text_value):
```
Comment (TanyaKalistruk) [inline methods.py:21]: good idea)

File: methods.py [hunk 3]
```diff
@@ -1,33 +1,38 @@
...
+        """Clears the element with number elem_number"""
+        WebDriverWait(self.driver, 10).until(EC.presence_of_element_located(locator))
+        elements = self.get_elements(locator)
+        elements[elem_number].clear()
+
+    def send_keys(self, locator, keys, elem_number=0):
+        """Send keys to the element with number elem_number"""
+        WebDriverWait(self.driver, 10).until(EC.presence_of_element_located(locator))
+        elements = self.get_elements(locator)
+        elements[elem_number].send_keys(keys)
```
Comment (TanyaKalistruk) [inline methods.py]: please, add one more empty line at the end of file

File: pages/header.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+from methods import Methods
+from locators import Locators
+
+class Header():
+    """This is parent class"""
```
Comment (TanyaKalistruk) [inline pages/header.py]: actually, it is not a parent page. it is just page for header

Comment (dbabyuk) [reply pages/header.py]: Ok, I'll change

File: pages/header.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    def __init__(self, driver):
+        self.driver = driver
+        self.methods = Methods(self.driver)
```
Comment (TanyaKalistruk) [inline pages/header.py]: just an idea: make an inheritance for this class from Methods class and just use `super().__init__(driver)` and use methods of this class without self.methods

File: pages/home_page.py [hunk 6]
```diff
@@ -1,17 +1,11 @@
 from methods import Methods
 from locators import Locators
+from pages.header import Header
 
-class HomePage:
+class HomePage(Header):
```
Comment (TanyaKalistruk) [inline pages/home_page.py]: do you use there something from Header class?

File: pages/user_profile.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+    def click_logout(self):
+        try:
+            self.click_dropdown(1)
+        except:
+            self.click_icon()
+            self.click_dropdown(1)
+
+
+    # def enter_user_data(self, key_id, data_entry):
```
Comment (TanyaKalistruk) [inline pages/user_profile.py]: do not commit commented code

File: test_login_out.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
+from init import BasePage
+import unittest
+from pages.home_page import HomePage
+from pages.sign_in_page import SignInPage
+
+
+from credentials import Credentials
+
+
+class Test1(BasePage):
```
Comment (TanyaKalistruk) [inline test_login_out.py]: please, rename class

File: pages/home_page.py (hunk 6 above)
Comment (dbabyuk) [reply pages/home_page.py]: yep. So far this HomePage is not completed. More methods will be added later

File: pages/user_profile.py (hunk 7 above)
Comment (dbabyuk) [reply pages/user_profile.py]: Ok, I won't

File: test_login_out.py (hunk 8 above)
Comment (dbabyuk) [reply test_login_out.py]: Ok

File: .idea/Task_tests.iml [hunk 9]
```diff
@@ -1,14 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (rvynnychuk) [inline .idea/Task_tests.iml:1]: Please add .iml, .pyc and .xml files to gitignore.

File: README.md [hunk 10]
```diff
@@ -1,5 +1,10 @@
 # Task_tests
```
Comment (rvynnychuk) [inline README.md:1]: Should contain steps for project setup too.

File: locators.py (hunk 1 above)
Comment (rvynnychuk) [reply locators.py]: Is this for login page? If so, name should be correspondent.

File: locators.py [hunk 11]
```diff
@@ -1,38 +1,30 @@
...
+    EMAIL = (By.ID, "username")
+    PASSWORD = (By.ID, "password")
+    SIGN_IN = (By.CSS_SELECTOR, "input[value='Sign In']")
 
+    SIGNUP = (By.ID, "login1")
+    PASSWORD_SIGNUP = (By.ID, "password1")
+    PASSWORD_VERIFY = (By.ID, "matchingPassword")
+    REGISTER_BTN = (By.CSS_SELECTOR, "input[value='Register']")
 
+class LocatorsUsers:
```
Comment (rvynnychuk) [inline locators.py]: Maybe `RegisterUserLocators`?

File: locators.py [hunk 12]
```diff
@@ -1,38 +1,30 @@
...
 
+     user_fields = {
```
Comment (rvynnychuk) [inline locators.py]: Do we need additional object for this?

File: methods.py [hunk 13]
```diff
@@ -1,33 +1,38 @@
-from selenium.webdriver.common.by import By
 from selenium.webdriver.support import expected_conditions as EC
```
Comment (rvynnychuk) [inline methods.py:1]: Maybe `webdriver_wrapper`?

Comment (rvynnychuk) [reply methods.py:1]: It is for file and class name.

File: methods.py [hunk 14]
```diff
@@ -1,33 +1,38 @@
...
-        element = self.get_element(locator_type, locator)
-        element.send_keys(keys)
-
-
-            
+    def get_elements(self, locator):
+        """Returns all elements for the specific locator"""
+        return self.driver.find_elements(*locator)
+
+    def click_element(self, locator, elem_number=0):
```
Comment (rvynnychuk) [inline methods.py:15]: Lets drop this method in favor of `click_element_by_text`

File: methods.py [hunk 15]
```diff
@@ -1,33 +1,38 @@
...
+        """Clicks on the element with number elem_number"""
+        WebDriverWait(self.driver, 10).until(EC.element_to_be_clickable(locator))
+        elements = self.get_elements(locator)
+        elements[elem_number].click()
+
+    def click_element_by_text(self, locator, text_value):
+        """Clicks on the element with text attribute text_value"""
+        WebDriverWait(self.driver, 10).until(EC.element_to_be_clickable(locator))
```
Comment (rvynnychuk) [inline methods.py]: Let's extract wait time into constant and place it in BasePage class.

File: methods.py [hunk 16]
```diff
@@ -1,33 +1,38 @@
...
+        elements = self.get_elements(locator)
+        for element in elements:
+            if element.text == text_value:
+                element.click()
+
+    def clear_element(self, locator, elem_number=0):
```
Comment (rvynnychuk) [inline methods.py]: Let's pass string param of field to be cleared, don't use indexes.

File: methods.py [hunk 17]
```diff
@@ -1,33 +1,38 @@
...
+        """Clears the element with number elem_number"""
+        WebDriverWait(self.driver, 10).until(EC.presence_of_element_located(locator))
+        elements = self.get_elements(locator)
+        elements[elem_number].clear()
+
+    def send_keys(self, locator, keys, elem_number=0):
```
Comment (rvynnychuk) [inline methods.py]: Same here.

File: pages/header.py [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+        self.locators = Locators
+
+    def click_icon(self):
+        """Clicks on the round icon"""
+        self.methods.click_element(self.locators.ICON)
+
+    def click_dropdown(self, pick_item):
```
Comment (rvynnychuk) [inline pages/header.py]: Looks like naming is wrong. click_dropdown -> select_option(self, option)

File: methods.py [hunk 19]
```diff
@@ -1,33 +1,38 @@
-from selenium.webdriver.common.by import By
 from selenium.webdriver.support import expected_conditions as EC
 from selenium.webdriver.support.wait import WebDriverWait
 
 
 class Methods(object):
```
Comment (rvynnychuk) [inline methods.py:5]: Let's create instance of this wrapper in BasePage class, and use it via inheritance.

File: pages/sign_in_page.py [hunk 20]
```diff
@@ -1,23 +1,23 @@
 from methods import Methods
 from locators import Locators
-from pages.home_page import HomePage
 
-class SignInPage(HomePage):
-    """Inherits click_icon method from HomePage"""
+
+class SignInPage():
+    """Sign in so far, Sign Up expected here as well"""
```
Comment (rvynnychuk) [inline pages/sign_in_page.py]: Avoid temporal docstrings.

File: pages/sign_in_page.py [hunk 21]
```diff
@@ -1,23 +1,23 @@
...
 
-    def clear_box(self, box_name):
-        self.methods.clear_element('id', self.locators.ID_LOC[box_name])
+    def clear_boxes(self):
+        self.methods.clear_element(self.locators.EMAIL)
+        self.methods.clear_element(self.locators.PASSWORD)
 
-    def send_keys(self, box_name, keys):
-        self.methods.send_keys('id', self.locators.ID_LOC[box_name], keys)
+    def enter_credentials(self, entry):
```
Comment (rvynnychuk) [inline pages/sign_in_page.py]: My advice is to use explicit params, username and password.

File: pages/user_profile.py [hunk 22]
```diff
@@ -0,0 +1,32 @@
+from methods import Methods
+from locators import Locators, LocatorsUsers
+from pages.home_page import HomePage
+
+class UserProfilePage(HomePage):
```
Comment (rvynnychuk) [inline pages/user_profile.py]: Why inheritance starts from HomePage, guess it should be BasePage.

File: pages/header.py (hunk 5 above)
Comment (dbabyuk) [reply pages/header.py]: Done. Commit amend is coming soon

File: locators.py (hunk 1 above)
Comment (dbabyuk) [reply locators.py]: Done

File: locators.py (hunk 11 above)
Comment (dbabyuk) [reply locators.py]: We got used to call them just Users (meaning registered users). Unregistered user we call Guest

File: locators.py (hunk 12 above)
Comment (dbabyuk) [reply locators.py]: It will be later needed for user profile page and of course this class of locators will be extended

File: locators.py (hunk 11 above)
Comment (rvynnychuk) [reply locators.py]: My point is, that you need better naming for this, tie it up to page somehow.

File: methods.py (hunk 19 above)
Comment (dbabyuk) [reply methods.py:5]: Ok, but Nazar has created BasePage so he has to participate in this

File: methods.py (hunk 14 above)
Comment (dbabyuk) [reply methods.py:15]: We can but later we will need it anyway

File: methods.py (hunk 15 above)
Comment (dbabyuk) [reply methods.py]: You mean TIMEOUT = 10 ?

File: methods.py (hunk 16 above)
Comment (dbabyuk) [reply methods.py]: Ok

File: methods.py (hunk 17 above)
Comment (dbabyuk) [reply methods.py]: Ok, I'll try

File: methods.py (hunk 3 above)
Comment (dbabyuk) [reply methods.py]: Done

File: pages/user_profile.py (hunk 22 above)
Comment (dbabyuk) [reply pages/user_profile.py]: Ok

File: pages/sign_in_page.py (hunk 20 above)
Comment (dbabyuk) [reply pages/sign_in_page.py]: Ok

File: methods.py (hunk 13 above)
Comment (dbabyuk) [reply methods.py:1]: Do you mean it will be inherited by other classes ?

File: README.md (hunk 10 above)
Comment (dbabyuk) [reply README.md:1]: in separate text file ?

File: locators.py (hunk 11 above)
Comment (dbabyuk) [reply locators.py]: Ok

File: pages/header.py (hunk 18 above)
Comment (dbabyuk) [reply pages/header.py]: good point

File: .idea/Task_tests.iml (hunk 9 above)
Comment (dbabyuk) [reply .idea/Task_tests.iml:1]: done

File: pages/sign_in_page.py (hunk 21 above)
Comment (dbabyuk) [reply pages/sign_in_page.py]: ???

File: README.md (hunk 10 above)
Comment (rvynnychuk) [reply README.md:1]: README.md is a good place for this

File: methods.py (hunk 15 above)
Comment (rvynnychuk) [reply methods.py]: Yeap

File: methods.py (hunk 14 above)
Comment (rvynnychuk) [reply methods.py:15]: Try to avoid redundant code now.

File: methods.py (hunk 19 above)
Comment (rvynnychuk) [reply methods.py:5]: Work together on this in this PR.

File: pages/sign_in_page.py (hunk 21 above)
Comment (rvynnychuk) [reply pages/sign_in_page.py]: entry -> username and password as string values

File: methods.py (hunk 16 above)
Comment (dbabyuk) [reply methods.py]: Done

File: methods.py (hunk 17 above)
Comment (dbabyuk) [reply methods.py]: Done

File: locators.py (hunk 12 above)
Comment (dbabyuk) [reply locators.py]: I removed everything about User Profile etc. It will be later added with other user stuff.

File: methods.py (hunk 19 above)
Comment (dbabyuk) [reply methods.py:5]: Donr

File: methods.py (hunk 14 above)
Comment (dbabyuk) [reply methods.py:15]: ok

**Pull request merged**